qiskit_ibm_runtime/ibm_runtime_service.py

```diff
@@ -843,13 +850,19 @@ def programs(
         """
         if skip is None:
             skip = 0
+        if name:
+            for program in list(self._programs.values())[skip : limit + skip]:
+                if program.name == name:
+                    self._programs[program.program_id] = program
+                else:
+                    del self._programs[program.program_id]
```

Comment:
1. We should loop through all programs in self._programs.
2. If we find a program with given name then we should add it to a temporary list.
3. After we find all programs with the given name, we should then apply limit and skip on the temporary list and immediately return it.

All this local filtering should happen only if refresh is not set to True and self._programs is non empty.

We shouldn’t delete programs from self._programs since it’s usually lot of info that’ll have to be refetched from sever.